### ISSUE TITLE
createdump fails on macOS 12 Monterey: Can not find the _dyld_all_image_infos symbol

### ISSUE THREAD

**Issue by joshuajung** [dotnet/runtime#62181]
### Description

Running `createdump` on macOS 12 Monterey (so far reproduced on M1/Apple Silicon) fails with

```
ERROR: Can not find the _dyld_all_image_infos symbol
```

I assume that this is related to `dotnet-dump` also not providing any output (see my original post in https://github.com/dotnet/diagnostics/issues/2518#issuecomment-974799706).

### Reproduction Steps

Run `createdump` from `/usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0`, e.g. like this:

```
./createdump 44185
```

### Expected behavior

A dump is created.

### Actual behavior

```
 ✘  /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0  ./createdump 44185
Gathering state for process 44185
 ✘  /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0  ./createdump 44185 -v
Gathering state for process 44185
0000000100990000 - 000000010099c000 (000003) 00000000     0 1 0 r-x 05
000000010099c000 - 00000001009a0000 (000001) 00000000     0 1 0 r-- 01
00000001009a0000 - 00000001009a4000 (000001) 00000000     0 1 0 rw- 03
[abbreviated]
0000600008000000 - 0000600010000000 (002000) 00000000     0 2 0 rw- 03
0000600010000000 - 0000600018000000 (002000) 00000000     0 3 0 rw- 03
0000600018000000 - 0000600020000000 (002000) 00000000     0 3 0 rw- 03
mach_vm_region_recurse for address 0000600020000000 08000000 FAILED 1 (os/kern) invalid address
TryFindDyLinker: found module header at 0000000100990000 0000c000 ncmds 18 sizeofcmds 00000818 type 02
TryFindDyLinker: found module header at 0000000100b18000 00060000 ncmds 17 sizeofcmds 00000730 type 07
TryFindDyLinker: found dylinker
CMD: load command cmd 19 (25) size 392
CMD: load bias 0000000100b18000
CMD: vmaddr 0000000000000000 vmsize 0000000000060000 fileoff 0000000000000000 filesize 0000000000060000 nsects 4 max r-x init r-x 00 __TEXT
     addr 0000000000001000 size 0000000000051510 off 00001000 align 0c flags 80000400 __text
     addr 0000000000052510 size 0000000000000ed0 off 00052510 align 04 flags 00 __const
     addr 00000000000533e0 size 0000000000009441 off 000533e0 align 00 flags 02 __cstring
     addr 000000000005c824 size 0000000000000334 off 0005c824 align 02 flags 00 __unwind_info
CMD: load command cmd 19 (25) size 312
CMD: vmaddr 0000000000060000 vmsize 0000000000010000 fileoff 0000000000060000 filesize 0000000000010000 nsects 3 max rw- init rw- 10 __DATA_CONST
     addr 0000000000060000 size 0000000000000070 off 00060000 align 03 flags 00 __auth_ptr
     addr 0000000000060070 size 00000000000041a8 off 00060070 align 03 flags 00 __const
     addr 0000000000064218 size 0000000000008000 off 00064218 align 00 flags 00 __bss
CMD: load command cmd 19 (25) size 472
CMD: vmaddr 0000000000070000 vmsize 0000000000004000 fileoff 0000000000070000 filesize 0000000000004000 nsects 5 max rw- init rw- 00 __DATA
     addr 0000000000070000 size 000000000000013c off 00070000 align 03 flags 00 __data
     addr 0000000000070140 size 0000000000000170 off 00070140 align 03 flags 00 __all_image_info__DATA
     addr 00000000000702b0 size 0000000000000040 off 000702b0 align 03 flags 00 __crash_info
     addr 0000000000070300 size 0000000000002208 off 00000000 align 06 flags 01 __common
     addr 0000000000072508 size 00000000000004f4 off 00000000 align 03 flags 01 __bss
CMD: load command cmd 19 (25) size 72
CMD: vmaddr 0000000000074000 vmsize 0000000000038000 fileoff 0000000000074000 filesize 0000000000034450 nsects 0 max r-- init r-- 00 __LINKEDIT
CMD: load command cmd 80000034 (-2147483596) size 16
CMD: load command cmd 80000033 (-2147483597) size 16
CMD: load command cmd 02 (2) size 24
CMD: load command cmd 0b (11) size 80
CMD: load command cmd 0f (15) size 32
CMD: load command cmd 1b (27) size 24
CMD: load command cmd 32 (50) size 32
CMD: load command cmd 2a (42) size 16
CMD: load command cmd 05 (5) size 288
CMD: load command cmd 1e (30) size 16
CMD: load command cmd 26 (38) size 16
CMD: load command cmd 29 (41) size 16
CMD: load command cmd 1d (29) size 16
SYM: symoff 0007b8f8 nsyms 2332 stroff 00084ab8 strsize 135224 iext 2331 next 1
ERROR: Can not find the _dyld_all_image_infos symbol
```

No dump is stored.

### Configuration

* .NET 6.0.100
* macOS Monterey 12.0.1 (21A559)
* ARM64

**Comment by mikem8361** [comment]
@joshuajung, when you ran createdump were you superuser? I ran `sudo -E ./createdump <pid>` against an app running 6.0.0 and it worked for me.

The other big difference is you are running Monterey 12.0.1 and I'm running version 11.6.

**Comment by joshuajung** [comment]
@mikem8361 Thanks for having a look at this. Yes, it also fails as super user, in the same way as pasted above. Unfortunately, I don't have a macOS 11 machine to triage further, but I will report in case future macOS updates have any effect.

**Comment by mikem8361** [comment]
On Monterey the _dyld_all_image_infos symbol is not an "external" symbol but part of the regular symbols. The problem is that there are 2332 regular symbols to read and linearly search instead of just 10 or so on Big Sur (one on Monterey). But I'll change the MachO reader that createdump uses to first go though the externs and then through all the rest of them.

**Comment by mikem8361** [comment]
This will be fixed in the next 6.0 service release (I think 6.0.3).
